Terminal.Gui/Views/SpinnerView.cs

```diff
@@ -1,44 +1,271 @@
+
+		private const SpinnerStyle DEFAULT_STYLE = SpinnerStyle.Line;
+		private const int DEFAULT_DELAY = 130;
+
+		private bool _hasSpecial = false;
+		private bool _bounceReverse = false;
+		private SpinnerStyle _style;
 		private int _currentIdx = 0;
 		private DateTime _lastRender = DateTime.MinValue;
 		private object _timeout;
 
+		/// <summary>
+		/// Gets or sets the frames used to animate the spinner.
+		/// </summary>
+		public string [] Frames { get; set; }
+
 		/// <summary>
 		/// Gets or sets the number of milliseconds to wait between characters
-		/// in the spin.  Defaults to 250.
+		/// in the spin.  Defaults to the SpinnerStyle's Interval value.
 		/// </summary>
 		/// <remarks>This is the maximum speed the spinner will rotate at.  You still need to
 		/// call <see cref="View.SetNeedsDisplay()"/> or <see cref="SpinnerView.AutoSpin"/> to
 		/// advance/start animation.</remarks>
-		public int SpinDelayInMilliseconds { get; set; } = 250;
+		public int SpinDelayInMilliseconds { get; set; } = DEFAULT_DELAY;
+
+		/// <summary>
+		/// Gets or sets whether spinner should go through frames in reverse order.
+		/// If SpinBounce is true, this sets the starting order.
+		/// </summary>
+		public bool SpinReverse { get; set; } = false;
+
+		/// <summary>
+		/// Gets or sets whether spinner should go back and forth through frames rather than
+		/// going to the end and starting again at the beginning.
+		/// </summary>
+		public bool SpinBounce { get; set; } = false;
+
+		/// <summary>
+		/// Gets the size in characters of the spinner.
+		/// </summary>
+		public int SpinnerWidth { get => GetSpinnerWidth (); }
+
+		/// <summary>
+		/// Gets whether the current spinner style contains only ASCII characters.
+		/// </summary>
+		public bool IsAsciiOnly { get => GetIsAsciiOnly (); }
+
+		/// <summary>
+		/// Gets whether the current spinner style contains emoji or other special characters.
+		/// </summary>
+		public bool HasSpecialCharacters { get => GetHasSpecial (); }
 
 		/// <summary>
 		/// Creates a new instance of the <see cref="SpinnerView"/> class.
 		/// </summary>
 		public SpinnerView ()
 		{
-			Width = 1; Height = 1;
+			SpinnerStyle = DEFAULT_STYLE;
+			Width = GetSpinnerWidth ();
+			Height = 1;
+		}
+
+		/// <summary>
+		/// Creates a new instance of the <see cref="SpinnerView"/> class.
+		/// </summary>
+		public SpinnerView (SpinnerStyle style)
+		{
+			SpinnerStyle = style;
+			Width = GetSpinnerWidth ();
+			Height = 1;
+		}
+
+		/// <summary>
+		/// Creates a new instance of the <see cref="SpinnerView"/> class with
+		/// a custom spinner.
+		/// </summary>
+		public SpinnerView (string [] frames, int delay = DEFAULT_DELAY)
+		{
+			if (frames is not null &&
+				frames.Length > 0) {
+				_style = SpinnerStyle.Custom;
+				Frames = frames;
+				Width = GetSpinnerWidth ();
+				SpinDelayInMilliseconds = delay;
+			} else {
+				Width = 1;
+			}
+			Height = 1;
+		}
+
+		private int GetSpinnerWidth ()
```

Comment:
Again, regarding `Width` - This raises a question for me:

What happens if `SpinnerWidth` is > `View.Width`?